Lib/asyncio/futures.py

```diff
@@ -353,10 +353,24 @@ def _copy_future_state(source, dest):
 
     The other Future may be a concurrent.futures.Future.
     """
-    assert source.done()
     if dest.cancelled():
         return
     assert not dest.done()
+
+    # Use _get_snapshot for futures that support it
+    if hasattr(source, '_get_snapshot'):
```

Comment:
asyncio always creates future via `concurrent.futures.Future` so this check isn't needed

Reply:
The doc string says: The other Future may be a concurrent.futures.Future, however AFAICT it always is.  I'll remove the fallback and always use the fast path

Reply:
I cannot find any path where this would copy an asyncio Future.  I will adjust the doc string and update the benchmark and opening text.

I was worried about `wrap_future` but it even has a check for this
```
def wrap_future(future, *, loop=None):
    """Wrap concurrent.futures.Future object."""
    if isfuture(future):
        return future
    assert isinstance(future, concurrent.futures.Future), \
        f'concurrent.futures.Future is expected, got {future!r}'
```

Reply:
There are tests that fail when removing it. However I can't see how it would ever be called internally or externally with an asyncio.Future

Reply:
AFAICT all asyncio futures go through `isfuture(future)` branch, not through `_copy_future_state`

Reply:
I have adjusted the tests as well but I did not adjust `_chain_future` to contain scope